compiler/optimizer/LoopCanonicalizer.cpp

```diff
@@ -2087,7 +2119,41 @@ void TR_LoopCanonicalizer::eliminateRedundantInductionVariablesFromLoop(TR_Regio
          }
       }
    }
+bool TR_LoopCanonicalizer::isTrivialArrayOperation(TR::NodeChecklist &visited,const TR::Node *node)
+   {
+      if (visited.contains(node))
+         return true;
 
+      for (auto childIndex = 0; childIndex < node->getNumChildren(); ++childIndex)
+         {
+         if (isTrivialArrayOperation(visited, node->getChild(childIndex)))
+            visited.add(node);
+         else
+            {
+            traceMsg(comp(),"LoopCanonicalizer.cpp::isTrivialArrayOperation Returning from for loop %s\n", node->getChild(childIndex)->getOpCode().getName());
+            return false;         
+            }
+         }
+
+      TR::ILOpCodes opCode = node->getOpCodeValue();
+      traceMsg(comp(),"LoopCanonicalizer.cpp::isTrivialArrayOperation opCode value %d\n",opCode);
+      if (TR::ILOpCodes::bstorei = opCode
+         && TR::ILOpCodes::aladd != opCode
+         && TR::ILOpCodes::aload != opCode
+         && TR::ILOpCodes::lsub != opCode
+         && TR::ILOpCodes::i2l != opCode
+         && TR::ILOpCodes::iload != opCode
+         && TR::ILOpCodes::lconst != opCode
+         && TR::ILOpCodes::i2b != opCode
+         && TR::ILOpCodes::iadd != opCode
+         && TR::ILOpCodes::istore != opCode
+         && TR::ILOpCodes::iconst != opCode)
```

Comment:
This seems too restrictive. What about trivial array operations on non-byte arrays? Those will not be ignored by this `if`. We should instead use the opcode APIs, for example `isStoreIndirect`, `isConversion`, `isAdd`, etc.

Reply:
Right, that's because I wanted to get the recognition working for the set of trees we are seeing in the benchmark first. But I'll definitely replace the opcode with APIs

Reply:
I think it would be better to use capability based queries so we can handle aladd and aidd, iadd and isub etc and make this less brittle across different modes. I know you want to keep it simple, but I agree this seems a bit overfit to the specific case you are looking at.